docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,436 @@
+is anticipated that the `kubelet` will parent containers based on quality of service
+in the future.  In that environment, those changes will be relative to the configured
+`cgroup-root`.
+
+### Linux Kernel Parameters
+
+The `kubelet` will set the following:
+
+* `sysctl -w vm.overcommit_memory=1`
+* `sysctl -w vm.panic_on_oom=0`
+* `sysctl -w kernel/panic=10`
+* `sysctl -w kernel/panic_on_oops=1`
+
+### OOM Score Adjustment
+
+The `kubelet` at bootstrapping will set the `oom_score_adj` value for Kubernetes
+daemons, and any dependent container-runtime daemons.
+
+If `container-runtime` is set to `docker`, then set its `oom_score_adj=-900`
+
+## Implementation concerns
+
+### kubelet block-level architecture
+
+```
++----------+       +----------+    +----------+
+|          |       |          |    | Pod      |
+|  Node    <-------+ Container<----+ Lifecycle|
+|  Manager |       | Manager  |    | Manager  |
+|          +------->          |    |          |
++---+------+       +-----+----+    +----------+
+    |                    |
+    |                    |
+    |  +-----------------+
+    |  |                 |
+    |  |                 |
++---v--v--+        +-----v----+
+| cgroups |        | container|
+| library |        | runtimes |
++---+-----+        +-----+----+
+    |                    |
+    |                    |
+    +---------+----------+
+              |
+              |
+  +-----------v-----------+
+  |     Linux Kernel      |
+  +-----------------------+
+```
+
+The `kubelet` should move to an architecture that resembles the above diagram:
+
+* The `kubelet` should not interface directly with the `cgroup` file-system, but instead
+should use a common `cgroups library` that has the proper abstraction in place to
+work with either `cgroupfs` or `systemd`.  The `kubelet` should just use `libcontainer`
+abstractions to facilitate this requirement.  The `libcontainer` abstractions as
+currently defined only support an `Apply(pid)` pattern, and we need to separate that
+abstraction to allow cgroup to be created and then later joined.
+* The existing `ContainerManager` should separate node bootstrapping into a separate
+`NodeManager` that is dependent on the configured `cgroup-driver`.
+* The `kubelet` flags for cgroup paths will convert internally as part of cgroup library,
+i.e. `/foo/bar` will just convert to `foo-bar.slice`
+
+### kubelet accounting for end-user pods
+
+This proposal re-enforces that it is inappropriate at this time to depend on `--cgroup-root` as the
+primary mechanism to distinguish and account for end-user pod compute resource usage.
+
+Instead, the `kubelet` can and should sum the usage of each running `pod` on the node to account for
+end-user pod usage separate from system-reserved and kubernetes-reserved accounting via `cAdvisor`.
+
+## Known issues
+
+### Docker runtime support for --cgroup-parent
+
+Docker versions <= 1.0.9 did not have proper support for `-cgroup-parent` flag on `systemd`.  This
```

Comment:
Docker 1.7 is minimal version Kubelet support since 1.1 release. Kubelet will mark the node not_ready today if docker version is below that minimal version.